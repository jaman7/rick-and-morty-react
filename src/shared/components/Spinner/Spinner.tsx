import React from 'react';
import { motion } from 'framer-motion';
import './Spinner.scss';

interface SpinnerProps {
  size?: 'small' | 'medium' | 'large';
  color?: string;
  ariaLabel?: string;
}

const Spinner: React.FC<SpinnerProps> = ({ size = 'medium', color = '#007bff', ariaLabel = 'Loading...' }) => {
  const sizes = {
    small: '20px',
    medium: '40px',
    large: '60px',
  };

  return (
    <div role="status" aria-label={ariaLabel} aria-busy="true">
      <motion.div
        className="loader"
        animate={{ rotate: 360 }}
        transition={{ repeat: Infinity, duration: 0.8, ease: 'linear' }}
        style={{
          width: sizes[size],
          height: sizes[size],
          border: `4px solid ${color}`,
          borderTopColor: 'transparent',
          borderRadius: '50%',
        }}
      />
      <span className="visually-hidden">{ariaLabel}</span>
    </div>
  );
};

export default Spinner;
