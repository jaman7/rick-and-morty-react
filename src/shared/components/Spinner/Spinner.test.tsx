import { render, screen } from '@testing-library/react';
import Spinner from './Spinner';
import { describe, it, expect } from 'vitest';

describe('Spinner', () => {
  it('renders spinner with default size and color', () => {
    render(<Spinner />);
    const spinnerWrapper = screen.getByRole('status');

    expect(spinnerWrapper).toBeInTheDocument();
    expect(spinnerWrapper).toHaveAttribute('aria-label', 'Loading...');
    expect(spinnerWrapper).toHaveAttribute('aria-busy', 'true');
  });

  it('renders spinner with custom size and color', () => {
    render(<Spinner size="large" color="#ff0000" />);
    const spinner = screen.getByRole('status').querySelector('.loader') as HTMLElement;

    expect(spinner).toBeInTheDocument();
    expect(spinner).toHaveStyle({
      width: '60px',
      height: '60px',
      border: '4px solid #ff0000',
    });

    expect(spinner.style.borderTopColor).toBe('transparent');
  });

  it('renders spinner with custom aria label', () => {
    render(<Spinner ariaLabel="Please wait" />);
    const spinnerWrapper = screen.getByRole('status');

    expect(spinnerWrapper).toHaveAttribute('aria-label', 'Please wait');
    expect(screen.getByText('Please wait')).toBeInTheDocument();
  });

  it('matches snapshot for default spinner', () => {
    const { container } = render(<Spinner />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('applies framer-motion infinite rotation animation (snapshot)', () => {
    const { asFragment } = render(<Spinner />);
    expect(asFragment()).toMatchSnapshot();
  });
});
