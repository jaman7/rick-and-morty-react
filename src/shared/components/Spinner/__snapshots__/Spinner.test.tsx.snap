// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spinner > applies framer-motion infinite rotation animation (snapshot) 1`] = `
<DocumentFragment>
  <div
    aria-busy="true"
    aria-label="Loading..."
    role="status"
  >
    <div
      class="loader"
      style="width: 40px; height: 40px; border: 4px solid rgb(0, 123, 255); border-top-color: transparent; border-radius: 50%;"
    />
    <span
      class="visually-hidden"
    >
      Loading...
    </span>
  </div>
</DocumentFragment>
`;

exports[`Spinner > matches snapshot for default spinner 1`] = `
<div
  aria-busy="true"
  aria-label="Loading..."
  role="status"
>
  <div
    class="loader"
    style="width: 40px; height: 40px; border: 4px solid rgb(0, 123, 255); border-top-color: transparent; border-radius: 50%;"
  />
  <span
    class="visually-hidden"
  >
    Loading...
  </span>
</div>
`;
